File: 3100-water-bottles-ii.py
class Solution:
    def maxBottlesDrunk(self, numBottles: int, numExchange: int) -> int:
        '''
        Each time we make an exchange, numExchange increases by 1
        '''
        bottlesDrank = 0
        emptyBottles = 0
        while numBottles + emptyBottles >= numExchange:
            bottlesDrank += numBottles
            emptyBottles += numBottles
            numBottles = 0
            while emptyBottles >= numExchange:
                emptyBottles -= numExchange
                numExchange += 1
                numBottles += 1

        return bottlesDrank + numBottles


sol = Solution()
print(f'output: {sol.maxBottlesDrunk(13, 6)}, expected: 15')
print(f'output: {sol.maxBottlesDrunk(10, 3)}, expected: 13')

File: 3100-water-bottles-ii.java
class WaterBottlesIISolution {

    public int maxBottlesDrunk(int numBottles, int numExchange) {
        int bottlesDrunk = 0;
        int emptyBottles = 0;
        while (numBottles + emptyBottles >= numExchange) {
            bottlesDrunk += numBottles;
            emptyBottles += numBottles;
            numBottles = 0;
            while (emptyBottles >= numExchange) {
                emptyBottles -= numExchange;
                numExchange++;
                numBottles++;
            }
        }

        return bottlesDrunk + numBottles;
    }

    public static void main(String[] args) {
        WaterBottlesIISolution sol = new WaterBottlesIISolution();
        System.out.println("Output: " + sol.maxBottlesDrunk(13, 6) + ", expected: 15");
        System.out.println("Output: " + sol.maxBottlesDrunk(10, 3) + ", expected: 13");
    }
}